A journey of 1000 miles starts with 1 step

alexdowad opened the pull request:
I have been doing a *lot* of cleanup on the Racc codebase lately; these are just a few commits which I picked out of my development branch to send for review first. Other, bigger changes will have to undergo thorough testing before I send them in a PR.

Feedback please!

tenderlove commented on .travis.yml:
```diff
@@ -3,7 +3,6 @@ sudo: false
 cache: bundler
 language: ruby
 rvm:
-  - 1.8.7
```
If we're going to drop 1.8 support, we should just drop all the way up to 2.2 and make this Racc 2.0.  Unfortunately I recently had to do a release for 1.8 users because apparently they still exist.

alexdowad replied:
Wow, fast response time, @tenderlove!

Maybe we can create a development branch for commits approved for inclusion in the future 2.0 release, and I will send you PRs for merging into that branch?

tenderlove replied:
¯\_(ツ)_/¯ 

We can just make master 2.0 and backport bug fixes to the 1-4-stable branch.  I have no problems making master 2.0.

tenderlove commented:
This looks fine to me (other than my comment about 1.8 support).  I realized there's test coverage, but I'm mostly worried that the state tables we generate aren't changed (or if they are that they have the same states / transitions but maybe not the same labels).

alexdowad commented:
> I'm mostly worried that the state tables we generate aren't changed

Please see the `test/regress.rb` script which I just added. This is a first, crude start at a regression test suite which will ensure that Racc's output does not change at all from the last `1.4` release. (Unless the change is deliberate and desired.)

I ran this on `ruby18.y` from `parser`, and it actually caught a change in output caused by one of my refactorings. I have adjusted that commit so that the output on `ruby18.y` is exactly the same as `master` now.

alexdowad commented:
I'm quite happy about this regression test script -- this will give me confidence to go ahead with even bigger refactorings.

alexdowad commented:
OK, 9 regression tests are now part of the main test suite, which is automatically run by Travis CI on each PR!

The 9 tests are based on 9 real-world grammars, contributed by various gem developers whose gems depend on Racc.

I have reached out to an additional 5 gem developers to see if they will contribute their grammars for more regression test cases.

alexdowad commented:
Besides catching one bug in this PR, and another bug in the next PR which I am currently preparing, the new regression test suite has just found another problem -- Racc's output is different on MRI 2.2.0 than it is on 1.9, 2.0, and 2.1!

alexdowad commented:
I've found the problem. It appears that the implementation of `Array#sort!` changed in 2.2.

alexdowad commented:
Hmm. Nobu made a commit on Sept 15, 2015, which makes Ruby use `qsort_r` as its quicksort implementation if it is available on the underlying platform. That could very well be the source of this difference.

The solution is to make the sort stable, probably by adding an additional sort key to break ties.

This will *not* produce the exact same results as `ruby_qsort` on MRI 2.1 and lower. So, Racc's output (state and transition tables, etc) will change on *both* 2.2 and 2.1-. However, the new output will be consistent between 2.1- and 2.2+.

alexdowad commented:
@tenderlove, can I go ahead and make that change?

alexdowad commented:
Oh, I forgot to tell you: the "problem" line is `statetransitiontable.rb:208`.

alexdowad commented:
I have rebased on #59, and added a 10th regression test. The "known good" output for all the regression tests is from #59. (The newly-consistent output for all versions of Ruby.)

alexdowad commented:
@mogigoma agreed to let the grammar he wrote for the `nasl` gem be used for another regression test case.

drbrain commented on test/regress/bibtex:
```diff
@@ -0,0 +1,460 @@
+##### State transition tables begin ###
+
+racc_action_table = [
+14, 33, 38, 26, 32, 26, 24, 18, 48, 15, 34, 33, 16, 37, 32, 
```
Is there a benefit to not pretty-printing these tables?

The old way makes them look slightly less intimidating

alexdowad replied:
Do you want this adjustment to be made before we go forward?

alexdowad replied:
> Is there a benefit to not pretty-printing these tables?

Absolutely not. This is what `master` did before I got to it.

drbrain replied:
If this is what master did then I have no issue with this difference.

drbrain commented:
I have no issues with this PR on the proprietary racc-based parser I wrote for work.

tenderlove commented:
> @tenderlove, can I go ahead and make that change?

Yes, go for it.  I think having the generated grammars checked in seems like a good way to prevent regressions for now.  In the future, I'd like to get away from it though.  It's more important that the tables always represent the same graph than have the same state labels, but I think that will be a bit harder to test.

alexdowad commented:
> It's more important that the tables always represent the same graph than have the same state labels

Sure, the tests can get smarter later. Once the big code cleanup is done, I think it will be a lot easier to see what we have and how to improve it. For now, these regression tests will give a high degree of confidence that the code cleanup isn't breaking anything.

alexdowad commented:
> Sure, the tests can get smarter later.

Let me add one thing here: before I attempt to touch the runtime, there will have to be more tests which provide thorough coverage of that. For now, I am just working on the parts which generate the parsing tables.

alexdowad commented:
@luke-gru agreed that his grammar for the `riml` gem can be used for another test case. Added.

A nice benefit of having an existing user base: you can scour the Internet for existing users and borrow their code for test cases. My feeling is that real-life usage will probably include all kinds of things which I would never think of if I was just coming up with test cases on my own.

alexdowad commented:
I've found another problem which we may want to fix, *before* the introduction of the regression test cases. You will find that the header of *every* Racc-generated parser says something like:

```ruby
#
# DO NOT MODIFY!!!!
# This file is automatically generated by Racc 1.4.13
# from Racc grammar file "".
#
```

Hmm... Racc grammar file ""?? Why didn't anybody notice that before???

By adding a single line of code to `grammarfileparser.rb`, I can make the actual grammar file name appear in those quote marks.

alexdowad commented:
Added another regression test case using a grammar file donated by @divoxx.

alexdowad commented:
Please note the new commit 06671a53, which has been added at the beginning of this PR. All of the regression tests are now based on this commit (the "known good" output is that produced by this commit).

alexdowad commented:
We have 16 regression test cases now, including grammars donated by @jamesotron and @uwabami. I have gone through and double-checked that all donated grammar files bear the original copyright notice and license (if any).

@tenderlove Anything else to do before this is merged?

alexdowad commented:
Added 2 more test cases based on grammars from @oggy and @whoward. I have ensured that all the added files appear in the manifest.

alexdowad commented:
Merged with an additional regression test based on a grammar donated by @adymo.
